wincrypto/src/dtls.rs

```diff
@@ -0,0 +1,601 @@
+
+        let schannel_cred = SCHANNEL_CRED {
+            dwVersion: SCHANNEL_CRED_VERSION,
+            hRootStore: windows::Win32::Security::Cryptography::HCERTSTORE(std::ptr::null_mut()),
+
+            grbitEnabledProtocols: if active {
+                SP_PROT_DTLS1_2_CLIENT
+            } else {
+                SP_PROT_DTLS1_2_SERVER
+            },
+
+            cCreds: cert_contexts.len() as u32,
+            paCred: cert_contexts.as_mut_ptr() as *mut *mut CERT_CONTEXT,
+
+            cMappers: 0,
+            aphMappers: std::ptr::null_mut(),
+
+            cSupportedAlgs: 0,
+            palgSupportedAlgs: std::ptr::null_mut(),
+
+            dwMinimumCipherStrength: 128,
+            dwMaximumCipherStrength: 256,
+            dwSessionLifespan: 0,
+            dwFlags: SCH_CRED_MANUAL_CRED_VALIDATION,
+            dwCredFormat: 0,
+        };
+
+        unsafe {
+            // These are the outputs of AcquireCredentialsHandleA
+            let mut cred_handle = SecHandle::default();
+            let mut creds_expiry: i64 = 0;
+            AcquireCredentialsHandleW(
+                None,
+                UNISP_NAME_W,
+                if active {
+                    SECPKG_CRED_OUTBOUND
+                } else {
+                    SECPKG_CRED_INBOUND
+                },
+                None,
+                Some(&schannel_cred as *const _ as *const std::ffi::c_void),
+                None,
+                None,
+                &mut cred_handle,
+                Some(&mut creds_expiry),
+            )
+            .expect("Failed to generate creds");
+
+            self.cred_handle = Some(cred_handle);
+        }
+
+        self.state = EstablishmentState::Handshaking;
+    }
+
+    pub fn handle_receive(
+        &mut self,
+        datagram: Option<&[u8]>,
+    ) -> Result<WinCryptoDtlsEvent, WinCryptoError> {
+        let state = self.state;
+        match state {
+            EstablishmentState::Established => {
+                if let Some(datagram) = datagram {
+                    self.process_packet(datagram)
+                } else {
+                    warn!("Unexpectedly asked to process no message!");
+                    Ok(WinCryptoDtlsEvent::None)
+                }
+            }
+            EstablishmentState::Handshaking => self.handshake(datagram),
+            EstablishmentState::Failed => {
+                Err(WinCryptoError("Handshake failed".to_string()).into())
+            }
+            EstablishmentState::Idle => {
+                Err(WinCryptoError("Handshake not initialized".to_string()).into())
+            }
+        }
+    }
+
+    pub fn pull_datagram(&mut self) -> Option<Vec<u8>> {
+        self.output.pop_front()
+    }
+
+    pub fn next_timeout(&mut self, now: Instant) -> Option<Instant> {
+        match self.state {
+            EstablishmentState::Idle | EstablishmentState::Handshaking => {
+                Some(now + Duration::from_millis(500))
```

Comment:
where does the 500ms come from?

Reply:
from the OpenSSL one. I'm not sure what the timeout is on the handshake, but I assume this ensures that we prod the handshake periodically.